src/QsCompiler/QirGeneration/QIR/Values.cs

```diff
@@ -23,10 +24,112 @@ internal class QirValues
+
+        internal IValue DefaultValue(ResolvedType type)
+        {
+            if (type.Resolution.IsInt)
+            {
+                var value = this.sharedState.Context.CreateConstant(0L);
+                return this.sharedState.Values.FromSimpleValue(value, type);
+            }
+            else if (type.Resolution.IsDouble)
+            {
+                var value = this.sharedState.Context.CreateConstant(0.0);
+                return this.sharedState.Values.FromSimpleValue(value, type);
+            }
+            else if (type.Resolution.IsBool)
+            {
+                var value = this.sharedState.Context.CreateConstant(false);
+                return this.sharedState.Values.FromSimpleValue(value, type);
+            }
+            else if (type.Resolution.IsPauli)
+            {
+                return this.CreatePauli(QsPauli.PauliI);
+            }
+            else if (type.Resolution.IsResult)
+            {
+                var getZero = this.sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.ResultGetZero);
+                var constant = this.sharedState.CurrentBuilder.Call(getZero);
+                return this.sharedState.Values.From(constant, type);
+            }
+            else if (type.Resolution.IsQubit)
+            {
+                var value = Constant.ConstPointerToNullFor(this.sharedState.Types.Qubit);
+                return this.sharedState.Values.From(value, type);
+            }
+            else if (type.Resolution.IsRange)
+            {
+                return this.sharedState.CreateRange(
+                    this.sharedState.Context.CreateConstant(0L),
+                    this.sharedState.Context.CreateConstant(1L),
+                    this.sharedState.Context.CreateConstant(-1L));
+            }
+            else if (type.Resolution is ResolvedTypeKind.TupleType ts)
+            {
+                var values = ts.Item.Select(this.DefaultValue).ToArray();
+                return this.sharedState.Values.CreateTuple(values, allocOnStack: this.sharedState.TargetQirProfile);
+            }
+            else if (type.Resolution is ResolvedTypeKind.UserDefinedType udt)
+            {
+                var elementTypes = this.sharedState.GetItemTypes(udt.Item.GetFullName());
+                var values = elementTypes.Select(this.DefaultValue).ToArray();
+                return this.sharedState.Values.CreateCustomType(udt.Item.GetFullName(), values, allocOnStack: this.sharedState.TargetQirProfile);
+            }
+
+            if (type.Resolution is ResolvedTypeKind.ArrayType itemType)
+            {
+                return this.sharedState.Values.CreateArray(itemType.Item, Array.Empty<IValue>(), allocOnStack: this.sharedState.TargetQirProfile);
+            }
+            else if (type.Resolution.IsFunction || type.Resolution.IsOperation)
+            {
+                // We can't simply set this to null, unless the reference and alias counting functions
```

Comment:
Both the QIR spec and the current QIR Runtime implementation treat null callables for the reference and alias counting functions as no-op, so perhaps this can indeed be simplified?

Reply:
I meant to also either completely disable any memory management for QIR targets, or (which I prefer since it is cleaner and needed for using the new data structures independently of the target capability) update the memory management to actually properly work even when somethings are stack allocated.